pkg/providers/snow/objects.go

```diff
@@ -0,0 +1,126 @@
+	}
+	capiCluster := CAPICluster(clusterSpec, snowCluster, kubeadmControlPlane)
+
+	return []runtime.Object{capiCluster, snowCluster, kubeadmControlPlane, controlPlaneMachineTemplate}, nil
+}
+
+func WorkersObjectsForCreate(clusterSpec *cluster.Spec) ([]runtime.Object, error) {
+	kubeadmConfigTemplates, err := KubeadmConfigTemplates(clusterSpec)
+	if err != nil {
+		return nil, err
+	}
+	workerMachineTemplates := SnowMachineTemplates(clusterSpec)
+	machineDeployments := MachineDeployments(clusterSpec, kubeadmConfigTemplates, workerMachineTemplates)
+
+	return concatWorkersObjects(machineDeployments, kubeadmConfigTemplates, workerMachineTemplates), nil
+}
+
+func ControlPlaneObjectsForUpgrade(ctx context.Context, oldSpec, newSpec *cluster.Spec, kubeClient KubeClient) ([]runtime.Object, error) {
+	snowCluster := SnowCluster(newSpec)
+	new := SnowMachineTemplate(newSpec.SnowMachineConfigs[newSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name])
+
+	old, err := oldControlPlaneMachineTemplate(ctx, kubeClient, oldSpec)
+	if err != nil {
+		return nil, err
+	}
+	if err := NewMachineTemplate(new, old); err != nil {
+		return nil, err
+	}
+
+	kubeadmControlPlane, err := KubeadmControlPlane(newSpec, new)
+	if err != nil {
+		return nil, err
+	}
+	capiCluster := CAPICluster(newSpec, snowCluster, kubeadmControlPlane)
+
+	return []runtime.Object{capiCluster, snowCluster, kubeadmControlPlane, new}, nil
+}
+
+func WorkersObjectsForUpgrade(ctx context.Context, oldSpec, newSpec *cluster.Spec, kubeClient KubeClient) ([]runtime.Object, error) {
+	kubeadmConfigTemplates, err := KubeadmConfigTemplates(newSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	workerMachineTemplates, err := WorkerMachineTemplates(ctx, kubeClient, oldSpec, newSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	machineDeployments := MachineDeployments(newSpec, kubeadmConfigTemplates, workerMachineTemplates)
+
+	return concatWorkersObjects(machineDeployments, kubeadmConfigTemplates, workerMachineTemplates), nil
+}
+
+func concatWorkersObjects(machineDeployments map[string]*clusterv1.MachineDeployment,
+	kubeadmConfigTemplates map[string]*v1beta1.KubeadmConfigTemplate,
+	workerMachineTemplates map[string]*snowv1.AWSSnowMachineTemplate,
+) []runtime.Object {
+	workersObjs := make([]runtime.Object, 0, len(machineDeployments)+len(kubeadmConfigTemplates)+len(workerMachineTemplates))
+	for _, item := range machineDeployments {
+		workersObjs = append(workersObjs, item)
+	}
+	for _, item := range kubeadmConfigTemplates {
+		workersObjs = append(workersObjs, item)
+	}
+	for _, item := range workerMachineTemplates {
+		workersObjs = append(workersObjs, item)
+	}
+	return workersObjs
+}
+
+func NewMachineTemplateName(new, old *snowv1.AWSSnowMachineTemplate) (string, error) {
+	if equality.Semantic.DeepDerivative(new.Spec, old.Spec) {
+		return old.GetName(), nil
+	}
+	return common.IncrementName(old.GetName())
+}
+
+func NewMachineTemplate(new, old *snowv1.AWSSnowMachineTemplate) error {
```

Comment:
What about something like
```suggestion
func UpdateMachineTemplateName(new, old *snowv1.AWSSnowMachineTemplate) error {
```